content/knowledge-base/guides/write-a-composition-function-in-go.md

```diff
@@ -0,0 +1,832 @@
+
+	in := &v1beta1.Input{}
+	if err := request.GetInput(req, in); err != nil {
+		response.Fatal(rsp, errors.Wrapf(err, "cannot get Function input from %T", req))
+		return rsp, nil
+	}
+
+	response.Normalf(rsp, "I was run with input %q", in.Example)
+	return rsp, nil
+}
+```
+
+All Go composition functions have a `RunFunction` method. Crossplane passes
+everything the function needs to run in a
+{{<hover label="hello-world" line="1">}}RunFunctionRequest{{</hover>}} struct.
+The function tells Crossplane what resources it should compose by returning a
+{{<hover label="hello-world" line="13">}}RunFunctionResponse{{</hover>}} struct.
+
+{{<hint "tip">}}
+Crossplane generates the `RunFunctionRequest` and `RunFunctionResponse` structs
+using [Protocol Buffers](http://protobuf.dev). You can find detailed schemas for
+`RunFunctionRequest` and `RunFunctionResponse` in the
+[Buf Schema Registry](https://buf.build/crossplane/crossplane/docs/main:apiextensions.fn.proto.v1beta1).
+{{</hint>}}
+
+Edit the `RunFunction` method to replace it with this code.
+
+```go
+func (f *Function) RunFunction(_ context.Context, req *fnv1beta1.RunFunctionRequest) (*fnv1beta1.RunFunctionResponse, error) {
+	rsp := response.To(req, response.DefaultTTL)
+
+	xr, err := request.GetObservedCompositeResource(req)
+	if err != nil {
+		response.Fatal(rsp, errors.Wrapf(err, "cannot get observed composite resource from %T", req))
+		return rsp, nil
+	}
+
+	region, err := xr.Resource.GetString("spec.region")
+	if err != nil {
+		response.Fatal(rsp, errors.Wrapf(err, "cannot read spec.region field of %s", xr.Resource.GetKind()))
+		return rsp, nil
+	}
+
+	names, err := xr.Resource.GetStringArray("spec.names")
+	if err != nil {
+		response.Fatal(rsp, errors.Wrapf(err, "cannot read spec.names field of %s", xr.Resource.GetKind()))
+		return rsp, nil
+	}
+
+	desired, err := request.GetDesiredComposedResources(req)
+	if err != nil {
+		response.Fatal(rsp, errors.Wrapf(err, "cannot get desired resources from %T", req))
+		return rsp, nil
+	}
+
+	_ = v1beta1.AddToScheme(composed.Scheme)
+
+	for _, name := range names {
+		b := &v1beta1.Bucket{
+			ObjectMeta: metav1.ObjectMeta{
+				Annotations: map[string]string{
+					"crossplane.io/external-name": name,
+				},
+			},
+			Spec: v1beta1.BucketSpec{
+				ForProvider: v1beta1.BucketParameters{
+					Region: ptr.To[string](region),
+				},
+			},
+		}
+
+		cd, err := composed.From(b)
+		if err != nil {
+			response.Fatal(rsp, errors.Wrapf(err, "cannot convert %T to %T", b, &composed.Unstructured{}))
+			return rsp, nil
+		}
+
+		desired[resource.Name("xbuckets-"+name)] = &resource.DesiredComposed{Resource: cd}
+	}
+
+	if err := response.SetDesiredComposedResources(rsp, desired); err != nil {
+		response.Fatal(rsp, errors.Wrapf(err, "cannot set desired composed resources in %T", rsp))
+		return rsp, nil
+	}
+
+	return rsp, nil
+}
+```
+
+Expand the below block to view the full `fn.go`, including imports and
```

Comment:
i think this is the right thing to do, to have this full code file hidden behind an expand accordion box, but I'm still having trouble visually noticing them when scanning a page and knowing to click/expand them. Are you having trouble with that too still?

Reply:
No, I don't have too much trouble with this. I think here it's a little better than in some cases, since the important bits of this code appear outside of an expand box.

Reply:
cool! your usage of text like `Expand the below block...` also helps mitigate this